### DISCUSSION TITLE
Is there an opportunity to send graphQL query in url query params?

### DISCUSSION THREAD

**Discussion by Bazalii** [quarkusio/quarkus#30759] [Q&A] [answered]
When I send a GET request `http://localhost:8080/graphql?query={getUser{id}}` I get 400 response code. This query works with standart POST request with a body. I don't understand if it is possible to send graphQL queries in query params but it is said in the [graphQL specification](https://graphql.org/learn/serving-over-http/ ) that _GraphQL HTTP server should handle the HTTP GET and POST methods_.

[a comment by an automated account is not shown]

**Reply by phillip-kruger** [reply]
You have to enable this. By default it's not enabled, for security reasons, as it expose details in the url. See [quarkus.smallrye-graphql.http.get.enabled](https://quarkus.io/guides/smallrye-graphql#quarkus-smallrye-graphql_quarkus.smallrye-graphql.http.get.enabled)

**Reply by Bazalii** [reply]
@phillip-kruger thank you, it works! But it is worth mentioning that syntax should be `query=%7BgetUser%7Bid%7D%7D` and not `query={getUser{id}}`.

I have another question which is not about the matter of this question but related to get requests to graphQL endpoint. Only when \*/\* is placed as first item in accept header or application/graphql+json is placed in accepted formats(priority doesn't matter here) request is processed correctly with 200 response code.

So, this works:
`Accept: */*,text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng;q=0.8,application/signed-exchange;v=b3;q=0.9`
`Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,application/graphql+json;q=0.8,application/signed-exchange;v=b3;q=0.9`

If regular header for google chrome is used, I get 406 error.

This doesn't work:
`Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9`
`Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*,image/avif,image/webp,image/apng;q=0.8,application/signed-exchange;v=b3;q=0.9`
`Accept: text/html,application/xhtml+xml,application/xml,*/*;q=0.9,image/avif,image/webp,image/apng;q=0.8,application/signed-exchange;v=b3;q=0.9`

Is it an expected behavior?

**Reply by phillip-kruger** [reply]
> But it is worth mentioning that syntax should be `query=%7BgetUser%7Bid%7D%7D` and not `query={getUser{id}}`.

This is normal url encoding that needs to happen, nothing specific to graphql. 

> I have another question which is not about the matter of this question but related to get requests to graphQL endpoint. Only when */* is placed as first item in accept header or application/graphql+json is placed in accepted formats(priority doesn't matter here) request is processed correctly with 200 response code.

The `*/*` order should not matter, so that might be a bug (@jmartisk) but in general it is better to add the accept header of the content type you want. As per https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#media-types

**Reply by Bazalii** [reply]
> This is normal url encoding that needs to happen, nothing specific to graphql.

Yes, but I think it is still worth mentioning here

> The `*/*` order should not matter, so that might be a bug (@jmartisk) but in general it is better to add the accept header of the content type you want. As per https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#media-types

I agree, but the case for usage of this "query in url" thing in my application is that user clicks the link and it is opened in standart browser which by default doesn't add this specific type of content to Accept header.

**Reply by jmartisk** [reply]
That we don't properly recognize `*/*` when it's not the first item, that's surely a bug. I've reported it as https://github.com/quarkusio/quarkus/issues/30806
